pex/resolver.py

```diff
@@ -167,420 +39,298 @@ def __new__(cls, requirement, distribution):
+  :type indexes: list of str
+  :keyword find_links: A list or URLs, paths to local html files or directory paths. If URLs or
+    local html file paths, these are parsed for links to distributions. If a local directory path,
+    its listing is used to discover distributons.
+  :type find_links: list of str
+  :keyword str cache: A directory path to use to cache distributions locally.
+  :keyword bool build: Whether to allow building source distributions when no wheel is found.
+    Defaults to ``True``.
+  :keyword bool use_wheel: Whether to allow resolution of pre-built wheel distributions.
+    Defaults to ``True``.
+  :keyword bool compile: Whether to pre-compile resolved distribution python sources.
+    Defaults to ``False``.
+  :yields: All :class:`ResolvedDistribution` instances meeting ``requirements`` for all
+    specifed interpreters and platforms.
   :raises Unsatisfiable: If ``requirements`` is not transitively satisfiable.
   :raises Untranslateable: If no compatible distributions could be acquired for
     a particular requirement.
   """
 
-  interpreters = interpreters or [PythonInterpreter.get()]
-  platforms = platforms or ['current']
+  curried_resolve = functools.partial(resolve,
+                                      requirements=requirements,
+                                      requirement_files=requirement_files,
+                                      constraint_files=constraint_files,
+                                      allow_prereleases=allow_prereleases,
+                                      transitive=transitive,
+                                      indexes=indexes,
+                                      find_links=find_links,
+                                      cache=cache,
+                                      build=build,
+                                      use_wheel=use_wheel,
+                                      compile=compile)
+
+  def iter_kwargs():
+    if not interpreters and not platforms:
+      yield dict(interpreter=None, platform=None)
+      return
+
+    if interpreters:
+      for interpreter in interpreters:
+        yield dict(interpreter=interpreter)
+
+    if platforms:
+      for platform in platforms:
+        yield dict(platform=platform)
 
   seen = set()
-  for interpreter in interpreters:
-    for platform in platforms:
-      for resolvable in resolve(requirements,
-                                fetchers,
-                                interpreter,
-                                platform,
-                                context,
-                                precedence,
-                                cache,
-                                cache_ttl,
-                                allow_prereleases,
-                                use_manylinux=use_manylinux,
-                                transitive=transitive):
-        if resolvable not in seen:
-          seen.add(resolvable)
-          yield resolvable
+  for kwargs in iter_kwargs():
```

Comment:
NB: The old cross-product loop was very broken. When a platform is specified, the resolve is foreign and we should definitley not try to use one or more local interpreters to resolve the foreign platform wheels. One bit of brokenness: In the past for a resolve with a local python2.7 interpreter and python3.6 interpreter across (inexact) platforms of linux-x86_64 and macosx-10.13-x86_64 we'd accept any abi for the foreign platform _per distribution_ which was broken in general and only worked when the indexes and find_links passed to resolve were carefully controlled to have uniform wheel platforms.